pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/BrokerCache.java

```diff
@@ -0,0 +1,121 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.Future;
+import org.apache.pinot.spi.utils.builder.ControllerRequestURLBuilder;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.BoundRequestBuilder;
+import org.asynchttpclient.Dsl;
+import org.asynchttpclient.Response;
+
+
+/**
+ * Maintains table -> list of brokers, supports update
+ * TODO can we introduce a simple websocket based controller endpoint to make the update reactive in the client?
+ */
+public class BrokerCache {
+
+  private static final TypeReference<Map<String, List<String>>> RESPONSE_TYPE_REF =
+      new TypeReference<Map<String, List<String>>>() {
+      };
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private final Random _random = new Random();
+  private final AsyncHttpClient _client;
+  private final String _address;
+  private volatile BrokerData _brokerData;
+
+  public BrokerCache(String scheme, String controllerHost, int controllerPort) {
+    _client = Dsl.asyncHttpClient();
+    ControllerRequestURLBuilder controllerRequestURLBuilder =
+        ControllerRequestURLBuilder.baseUrl(scheme + "://" + controllerHost + ":" + controllerPort);
+    _address = controllerRequestURLBuilder.forLiveBrokerTablesGet();
+  }
+
+  protected void updateBrokerData()
+      throws Exception {
+    BoundRequestBuilder getRequest = _client.prepareGet(_address);
+    Future<Response> responseFuture = getRequest.addHeader("accept", "application/json").execute();
+    Response response = responseFuture.get();
+    String responseBody = response.getResponseBody(StandardCharsets.UTF_8);
+    Map<String, List<String>> responses = OBJECT_MAPPER.readValue(responseBody, RESPONSE_TYPE_REF);
+    List<String> brokers = new ArrayList<>();
+    Map<String, List<String>> tableToBrokersMap = new HashMap<>();
+    Set<String> uniqueTableNames = new HashSet<>();
+    for (Map.Entry<String, List<String>> tableToBrokers : responses.entrySet()) {
+      List<String> brokersForTable = new ArrayList<>();
+      tableToBrokers.getValue().forEach(br -> {
+        String[] split = br.split("_");
+        String brokerHostPort = split[1] + ":" + split[2];
+        brokersForTable.add(brokerHostPort);
+        if (!brokers.contains(brokerHostPort)) {
+          brokers.add(brokerHostPort);
+        }
+      });
+      String tableName = tableToBrokers.getKey();
+      tableToBrokersMap.put(tableName, brokersForTable);
+
+      String tableNameWithoutSuffix =
+          tableName.replace(ExternalViewReader.OFFLINE_SUFFIX, "").replace(ExternalViewReader.REALTIME_SUFFIX, "");
+      uniqueTableNames.add(tableNameWithoutSuffix);
+    }
+
+    // Add table names without suffixes
+    uniqueTableNames.forEach(tableName -> {
+      if (!tableToBrokersMap.containsKey(tableName)) {
+        List<String> unifiedList =
```

Comment:
We want to do a intersection instead of a union (this is not even union, but combining 2 lists)